pkg/controller.go

```diff
@@ -127,46 +144,101 @@ func (c *controller) Run(ctx context.Context) error {
 	}
 }
 
-func (c *controller) evictPodsCloseToMemoryLimit(ctx context.Context) error {
-	podList, err := c.listPodsCloseToMemoryLimit(ctx)
-	if err != nil {
-		return fmt.Errorf("error listing Pods to evict: %w", err)
-	}
-	for _, pod := range podList {
-		evictionPolicy := policyv1beta1.Eviction{
-			ObjectMeta: pod.ObjectMeta,
-		}
-		if c.opts.DryRun {
-			evictionPolicy.DeleteOptions = &metav1.DeleteOptions{
-				DryRun: []string{"All"},
-			}
-		}
-		klog.V(1).Infof("Evicting Pod '%s/%s'", pod.Namespace, pod.Name)
-		c.recorder.Event(pod.DeepCopyObject(), "Normal", "SoftEviction", fmt.Sprintf("Pod '%s/%s' has at least one container close to its memory limit", pod.Namespace, pod.Name))
-
-		err := c.clientset.PolicyV1beta1().Evictions(pod.Namespace).Evict(ctx, &evictionPolicy)
+// evict pods with a pause between each eviction
+func (c *controller) evictWithPauseChanLoop(ctx context.Context) {
+	for pod := range c.pauseChan {
+		err := c.evictPod(ctx, pod)
 		if err != nil {
 			c.recorder.Event(pod.DeepCopyObject(), "Warning", "SoftEviction", fmt.Sprintf("Unable to evict Pod '%s/%s' : %v", pod.Namespace, pod.Name, err))
 			klog.Errorf("error evicting '%s/%s': %v", pod.Namespace, pod.Name, err)
 			continue
 		}
 
 		if c.opts.EvictionPause > 0 {
+			klog.V(2).Infof("Pausing pod eviction for %s", c.opts.EvictionPause)
 			time.Sleep(c.opts.EvictionPause)
 		}
+
+		klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
 	}
-	return nil
 }
 
-func (c *controller) listPodsCloseToMemoryLimit(ctx context.Context) ([]*corev1.Pod, error) {
+// evict pods having a PodDisruptionBudget. If the PDB is exceeded, the eviction
+// is retried after a delay using a different channel.
+func (c *controller) evictWithPDBChanLoop(ctx context.Context) {
+	backoffChan := make(chan *corev1.Pod, c.opts.ChannelBufferSize)
+	defer close(backoffChan)
+	go func() {
+		for pod := range backoffChan {
+			attempt := 1
+		retry:
+			klog.V(2).Infof("Retrying eviction of '%s/%s' (attempt %d)", pod.Namespace, pod.Name, attempt)
+			err := c.evictPod(ctx, pod)
+			if err != nil {
+				if strings.Contains(err.Error(), "disruption budget") && attempt <= 3 {
+					time.Sleep(time.Duration(math.Pow(5, float64(attempt))) * time.Second)
+					attempt++
+					goto retry
+				}
+				c.recorder.Event(pod.DeepCopyObject(), "Warning", "SoftEviction", fmt.Sprintf("Unable to evict Pod '%s/%s' : %v", pod.Namespace, pod.Name, err))
+				klog.Errorf("error evicting '%s/%s': %v", pod.Namespace, pod.Name, err)
+				continue
+			}
+
+			klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
+		}
+	}()
+
+	for pod := range c.pdbChan {
+		err := c.evictPod(ctx, pod)
```

Comment:
Since we don't observe any eviction pause for Pods with a PodDisruptionBudget, we could theoretically evict 10 at a time if the eviction was instantaneous

Given that, why not always start a goroutine to evict such Pods and be as fast as possible ? There is the risk we blow up the controller with long running evictions, but in theory, the PodDisruptionBudget should protect us from evicting all Pods of a same Deployment at once and disrupting a service.

Reply:
It goes with all at a time until it hits one, that it cannot evict due to the disruption budget being exhausted. This one is being retried and then blocks the queue from further execution.

I think it makes sense to remove the whole backoff logic and solely rely on the check interval. Will do that.

I was trying to evict all pods beyond the threshold as soon as possible. Maybe we will find a way to do this at a later point, but until now this is already an improvement over the status quo.